cpp/arcticdb/processing/aggregation.cpp

```diff
@@ -489,4 +509,157 @@ SegmentInMemory CountAggregatorData::finalize(const ColumnName& output_column_na
     return res;
 }
 
+/***********************
+ * FirstAggregatorData *
+ ***********************/
+
+void FirstAggregatorData::add_data_type(DataType data_type) {
+    add_data_type_impl(data_type, data_type_);
+}
+
+void FirstAggregatorData::aggregate(const std::optional<ColumnWithStrings>& input_column, const std::vector<size_t>& groups, size_t unique_values) {
+    if(data_type_.has_value() && *data_type_ != DataType::EMPTYVAL && input_column.has_value()) {
+        entity::details::visit_type(*data_type_, [&input_column, unique_values, &groups, this] (auto global_type_desc_tag) {
+            using GlobalInputType = decltype(global_type_desc_tag);
+            using GlobalTypeDescriptorTag =  typename OutputType<GlobalInputType>::type;
+            using GlobalRawType = typename GlobalTypeDescriptorTag::DataTypeTag::raw_type;
+            aggregated_.resize(sizeof(GlobalRawType)* unique_values);
+            auto col_data = input_column->column_->data();
+            auto out_ptr = reinterpret_cast<GlobalRawType*>(aggregated_.data());
+            entity::details::visit_type(input_column->column_->type().data_type(), [this, &groups, &out_ptr, &col_data] (auto type_desc_tag) {
+                using ColumnTagType = std::decay_t<decltype(type_desc_tag)>;
+                using ColumnType =  typename ColumnTagType::raw_type;
+                auto groups_pos = 0;
+                auto grp_idx = groups[0];
+                while (auto block = col_data.next<TypeDescriptorTag<ColumnTagType, DimensionTag<entity::Dimension::Dim0>>>()) {
+                    auto ptr = reinterpret_cast<const ColumnType *>(block.value().data());
+                    for (auto i = 0u; i < block.value().row_count(); ++i, ++ptr) {
+                        auto& val = out_ptr[groups[groups_pos]];
+                        bool is_first_group_el = (groups_cache_.find(groups[groups_pos]) == groups_cache_.end());
+                        if constexpr(std::is_floating_point_v<ColumnType>) {
+                            const auto& curr = GlobalRawType(*ptr);
+                            if ((grp_idx == groups[groups_pos]) && (is_first_group_el || std::isnan(static_cast<ColumnType>(val)))) {
```

Comment:
This logic looks like the output of will contain NaNs only if that is the only value in the group, is that correct? And is that the same behaviour as pandas?

Reply:
> This logic looks like the output of will contain NaNs only if that is the only value in the group, is that correct?

Exactly.

> And is that the same behaviour as pandas?

Yes.